zcash_primitives/src/sapling/circuit.rs

```diff
@@ -551,6 +553,82 @@ impl Circuit<bls12_381::Scalar> for Output {
     }
 }
 
+/// The parameters for the Sapling Spend circuit.
+pub struct SpendParameters(pub(crate) groth16::Parameters<Bls12>);
+
+impl SpendParameters {
+    /// Reads the parameters from their encoding.
+    ///
+    /// Only set `verify_point_encodings` to false if you are verifying the parameters in
+    /// another way (such as checking the hash of the parameters file on disk).
+    pub fn read<R: io::Read>(reader: R, verify_point_encodings: bool) -> io::Result<Self> {
+        groth16::Parameters::<Bls12>::read(reader, verify_point_encodings).map(Self)
+    }
+
+    /// Returns the verifying key for the Sapling Spend circuit.
+    pub fn verifying_key(&self) -> SpendVerifyingKey {
+        SpendVerifyingKey(self.0.vk.clone())
+    }
+
+    /// Returns the verifying key for the Sapling Spend circuit, with precomputations
+    /// optimized for verifying individual proofs.
+    pub fn prepared_verifying_key(&self) -> PreparedSpendVerifyingKey {
+        PreparedSpendVerifyingKey(groth16::prepare_verifying_key(&self.0.vk))
+    }
+}
+
+/// The verifying key for the Sapling Spend circuit.
+pub struct SpendVerifyingKey(pub(crate) groth16::VerifyingKey<Bls12>);
+
+impl SpendVerifyingKey {
+    /// Performs precomputations optimized for verifying individual proofs.
+    pub fn prepare(&self) -> PreparedSpendVerifyingKey {
+        PreparedSpendVerifyingKey(groth16::prepare_verifying_key(&self.0))
+    }
+}
+
+/// The verifying key for the Sapling Spend circuit, with precomputations optimized for
+/// verifying individual proofs.
+pub struct PreparedSpendVerifyingKey(pub(crate) groth16::PreparedVerifyingKey<Bls12>);
+
+/// The parameters for the Sapling Output circuit.
+pub struct OutputParameters(pub(crate) groth16::Parameters<Bls12>);
+
+impl OutputParameters {
+    /// Reads the parameters from their encoding.
+    ///
+    /// Only set `verify_point_encodings` to false if you are verifying the parameters in
+    /// another way (such as checking the hash of the parameters file on disk).
```

Comment:
`verify_point_encodings` just does on-curve and field element canonicity checks, right? Those should be really cheap; it's point decompression that is expensive (because it requires a square root).

Reply:
`bellman::groth16::Parameters` uses uncompressed encodings.

Reply:
Exactly, so `verify_point_encodings` is always cheap because it only checks field element canonicity and the curve equation ($-u^2 + v^2 = 1 + d_{\mathbb{J}} \cdot u^2 \cdot v^2$). Have we measured the overhead for the Sapling spend and output parameters if we just always set it to `true`?

Reply:
The curve equation in this case is for BLS12-381, as these are the proving system parameters.

The checks that `G1Affine::from_uncompressed` does (in constant time) are `is_on_curve` and `is_torsion_free`. The latter is more expensive because of multiplying the point by $x^2$.

`G2Affine::from_uncompressed` does the same checks. This time the point is only multiplied by $x$ in `is_torsion_free`, but the rest of the field arithmetic is more expensive (being in `Fp2`).

I haven't benchmarked the difference, but I presume @ebfull did when originally implementing it.

Reply:
Oh of course, sorry. Yeah that's a significant cost.